### PR TITLE
PLANNER-2759 Switch the kubernetes quickstart to ActiveMQ

### PR DESCRIPTION
**Pull request by rsynek**
- Switches from Kafka to ActiveMQ
- Shows off dynamic scaling

### JIRA
https://issues.redhat.com/browse/PLANNER-2759

### Referenced pull requests
https://github.com/kiegroup/optaplanner/pull/2117

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] tests</b>

- for a <b>full downstream build</b>  
  please add the label `run_fdb`

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] quarkus-branch</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] native</b>

- for <b>mandrel checks</b>  
  Run native checks against Mandrel image
  Please add comment: <b>Jenkins run mandrel</b>

- for a <b>specific mandrel check</b>  
  Run native checks against Mandrel image  
  Please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] mandrel</b>
</details>

### CODE REVIEW

File: technology/kubernetes/README.adoc [hunk 1]
```diff
@@ -19,20 +19,45 @@ to be able to interact with the OpenShift cluster.
...
+After the ArtemisCloud operator has been deployed, https://artemiscloud.io/docs/help/operator/#creating-operator-based-broker-deployments[create a new broker].
+The ArtemisCloud distribution comes with several `activemqartemis` example resources you can use as a starting point.
+
+This demo uses the AMQP protocol to send and receive messages.
+To enable the AMQP protocol on the broker, add the following acceptor definition in the `activemqartemis` resource `spec`:
+
+[source, yaml]
+----
+...
+spec:
```
Comment (rsynek) [inline technology/kubernetes/README.adoc:45]: I think we could omit this part if we used the Artemis core protocol, as there is a core acceptor enabled by default.

The downside is that the demo app would have to migrate from Smallrye messaging to JMS as well.

File: technology/kubernetes/demo-app/src/main/java/org/acme/demoapp/rest/DemoDataResource.java [hunk 2]
```diff
@@ -66,9 +74,16 @@ public Dataset createTimeTable(@PathParam("lessons") int lessons) {
             throw new IllegalArgumentException("The number of lessons (" + lessons + ") must be between 10 and 200.");
         }
         Dataset dataset = generateAndPersistUnsolvedDataset(lessons);
-        SolverEvent solverEvent = new SolverEvent(dataset.getProblemId(), SolverEventType.SOLVER_REQUEST);
+        SolverRequest solverRequest = new SolverRequest(dataset.getProblemId());
+        String solverRequestMessage;
+        try {
+            solverRequestMessage = objectMapper.writeValueAsString(solverRequest);
```
Comment (rsynek) [inline technology/kubernetes/demo-app/src/main/java/org/acme/demoapp/rest/DemoDataResource.java:80]: The amqp connector does not support Jackson OOTB.

File: technology/kubernetes/demo-app/pom.xml [hunk 3]
```diff
@@ -36,7 +36,7 @@
     </dependency>
     <dependency>
       <groupId>io.quarkus</groupId>
-      <artifactId>quarkus-smallrye-reactive-messaging-kafka</artifactId>
+      <artifactId>quarkus-smallrye-reactive-messaging-amqp</artifactId>
```
Comment (rsynek) [inline technology/kubernetes/demo-app/pom.xml:39]: The demo-app still uses Smallrye messaging, as it only sends requests and receives results. No scaling or long message processing is required.

File: technology/kubernetes/docker-compose.yml [hunk 4]
```diff
@@ -12,29 +12,14 @@ services:
...
       POSTGRES_DB: school-timetabling
 
-  zookeeper:
-    image: quay.io/strimzi/kafka:0.28.0-kafka-3.1.0
-    command: [
-      "sh", "-c",
-      "bin/zookeeper-server-start.sh config/zookeeper.properties"
-    ]
+  artemis:
+    image: quay.io/artemiscloud/activemq-artemis-broker:latest
```
Comment (rsynek) [inline technology/kubernetes/docker-compose.yml:16]: The same image the ArtemisCloud operator uses to spin up the broker.

File: technology/kubernetes/school-timetabling/src/test/resources/broker.xml [hunk 5]
```diff
@@ -0,0 +1,101 @@
+<?xml version='1.0'?>
```
Comment (rsynek) [inline technology/kubernetes/school-timetabling/src/test/resources/broker.xml:1]: Required by the `EmbeddedActiveMQ` in tests.

File: technology/kubernetes/README.adoc [hunk 6]
```diff
@@ -19,20 +19,45 @@ to be able to interact with the OpenShift cluster.
...
+  acceptors:
+  - name: amqp
+    connectionsAllowed: 100 <1>
```
Comment (triceo) [inline technology/kubernetes/README.adoc]: I have to say I do not like these references. If you copy-paste this code, the reference remains there, and that is bound to create problems for the less attentive users.

File: technology/kubernetes/common/src/main/java/org/acme/common/message/SolverRequest.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package org.acme.common.message;
+
+import io.quarkus.runtime.annotations.RegisterForReflection;
+
+@RegisterForReflection
```
Comment (triceo) [inline technology/kubernetes/common/src/main/java/org/acme/common/message/SolverRequest.java:5]: Not particularly excited about this. Who reflects over this?
Wouldn't it rather make sense we make this object properly serializable to JSON?

File: technology/kubernetes/school-timetabling/src/main/resources/application.properties [hunk 8]
```diff
@@ -37,17 +37,15 @@ quarkus.datasource.password=pgsql
...
-
-mp.messaging.outgoing.solver_out.topic=${solver.message.output}
-mp.messaging.outgoing.solver_out.value.serializer=io.quarkus.kafka.client.serialization.ObjectMapperSerializer
+solver.message.amq.host=localhost
+solver.message.amq.port=5672
+solver.message.amq.username=admin
+solver.message.amq.password=admin
+quarkus.qpid-jms.url=amqp://${solver.message.amq.host}:${solver.message.amq.port}?jms.prefetchPolicy.queuePrefetch=0
+quarkus.qpid-jms.username=${solver.message.amq.username}
+quarkus.qpid-jms.password=${solver.message.amq.password}
```
Comment (triceo) [inline technology/kubernetes/school-timetabling/src/main/resources/application.properties:50]: Doesn't this violate some sort of "thou shalt not put passwords in plain text" commandment?

**Review by triceo — APPROVED** [review]
Some comments, but overall, I don't see why not.

File: technology/kubernetes/README.adoc (hunk 6 above)
Comment (rsynek) [reply technology/kubernetes/README.adoc]: I agree it's not necessary for this code excerpt.

Comment (triceo) [reply technology/kubernetes/README.adoc]: It's in others too. I wonder can we instead use line numbers or something?

File: technology/kubernetes/common/src/main/java/org/acme/common/message/SolverRequest.java (hunk 7 above)
Comment (rsynek) [reply technology/kubernetes/common/src/main/java/org/acme/common/message/SolverRequest.java:5]: Jackson. Not sure if reflection can be avoided there. For instance, the Quarkus REST extension automatically registers the DTO classes for reflection [1].

[1] https://quarkus.io/guides/rest-json#about-serialization

File: technology/kubernetes/school-timetabling/src/main/resources/application.properties (hunk 8 above)
Comment (rsynek) [reply technology/kubernetes/school-timetabling/src/main/resources/application.properties:50]: The default value is right there, "admin:admin". When running in the OpenShift, the `solver.message.amq.username` and `solver.message.amq.password` (and some other properties too) get overridden by environment variables.

There is a convention that an environment variable `SOME_NICE_VARIABLE` gets converted to a property `some.nice.variable`.

File: technology/kubernetes/README.adoc (hunk 6 above)
Comment (rsynek) [reply technology/kubernetes/README.adoc]: Hmm, if we turn on line numbers, that could work. Let me try it.

File: technology/kubernetes/school-timetabling/src/main/resources/application.properties (hunk 8 above)
Comment (triceo) [reply technology/kubernetes/school-timetabling/src/main/resources/application.properties:50]: Yeah, but... *should* the password be right there?

Comment (rsynek) [reply technology/kubernetes/school-timetabling/src/main/resources/application.properties:50]: The default values are provided to make running the quickstart locally (no OpenShift, just docker-compose) easier.
Option A) remove it and the user has to provide the password via `-Dsolver.message.amq.password` when running the app
Option B) add a warning that the password should never be stored in this file in the production environment.

Comment (rsynek) [reply technology/kubernetes/school-timetabling/src/main/resources/application.properties:50]: Took option B).

**Pull request merged**